### ISSUE TITLE
Angular Element is not removed and readded to the DOM properly

### ISSUE THREAD

**Issue by yefim** [angular/angular#38778]
# 🐞 bug report

### Affected Package

The issue is caused by package @angular/elements

### Is this a regression?

No

### Description

When an Angular Element is detached from the DOM and then re-attached, none of the event bindings work (see minimal reproduction for demo). I am a Google engineer and this is causing issues for us (internal bug filed here: http://b/168062117).

I talked to @kseamon about this and he thinks the issue is located in either of these 2 places:

https://github.com/angular/angular/blob/a69507a0adc0e559d152eb3fb988a410ef9352a7/packages/elements/src/create-custom-element.ts#L219

https://github.com/angular/angular/blob/a69507a0adc0e559d152eb3fb988a410ef9352a7/packages/elements/src/component-factory-strategy.ts#L109-L125

Specifically, `this.scheduledDestroyFn` in not set to `null` in the scheduled destroy and so may be called again upon connect. I'm testing adding one line to see if it fixes the immediate issue.

However, even with this fix, there will still exist a tiny race condition that occurs when an element is scheduled to be destroyed and then connects before the destroy executes.

## 🔬 Minimal Reproduction

https://stackblitz.com/edit/angular-ovchwr?file=src%2Findex.html

## 🔥 Exception or Error

No user-visible exception

## 🌍  Your Environment

**Angular Version:**

I'm running it against the latest Angular in google3.

**Comment by gkalpak** [comment]
You are right, @yefim/@kseamon. It seems that by not setting `this.scheduledDestroyFn` to `null` when the component is destroyed, we hit [this branch][1] the next time the element is inserted into the DOM (and exit early, without instantiating the component).

I am not clear what you meant when you said "there will still exist a tiny race condition". Can you elaborate?

[1]: https://github.com/angular/angular/blob/a69507a0adc0e559d152eb3fb988a410ef9352a7/packages/elements/src/component-factory-strategy.ts#L93-L97

**Comment by gkalpak** [comment]
You are right, @yefim / @kseamon. It seems that by not setting `this.scheduledDestroyFn` to `null` when the component is destroyed, we hit [this branch][1] the next time the element is inserted into the DOM (and exit early, without instantiating the component).

I am not clear what you meant when you said "there will still exist a tiny race condition". Can you elaborate?

[1]: https://github.com/angular/angular/blob/a69507a0adc0e559d152eb3fb988a410ef9352a7/packages/elements/src/component-factory-strategy.ts#L93-L97

**Comment by gkalpak** [comment]
Tomake sure we are on the same page, here is what the intention is with the current implementation:
- When the custom element is removed from the DOM, the underlying component is scheduled for distruction after a short timeout (currently 10ms).
- If the custom element is inserted back into the DOM _before_ the timeout expires, then the component destruction is canceled and we keep the same component instance.
- If the custom element is inserted back into the DOM _after_ the timeout, then the component has been destroyed and we should instantiate a new underlying component.

This last bullet point is what we are failing to do at the moment, as mentioned in the previous comment (which is a bug), but that was the intention.

The reason we destroy the component instance is that there is some Angular-specific cleaning up that needs to happen when the element (and thus the underlying component instance) is no longer necessary (i.e. removed from the DOM for good).

The reason we have the short delay between removing the custom element from the DOM and destroying the component instance is to account for cases where you want to "move" the element around (i.e. remove it from its current location in the DOM and immediately insert at a new location). In that case, it would be a waste to destroy the component synchronously, only to re-instantiate it right after.

##
From a quick look, it seems that fixing the current bug (i.e. allowing the component to be re-instantiated when the host element is re-inserted into the DOM) is not trivial. For example, how do we handle [content projection][1]. The original nodes have been extracted from the host element the first time, so they are no longer there for us to extract. (We could cache them and re-use the same nodes, but I wouldn't be surprized if this broke in some edge cases.)

Even if we managed to fix this and allow the component to be re-instantiated, the internal state of the original component instance would not have been retained. This would lead to UI differences.

##
The main problem here is that there is no good way to automatically infer whether a removed element is intended to be re-used some time in the future (and thus the underlying component instance should be retained) or whether it is not needed any more (and thus the underlying component instance should be destroyed).

Therefore, we have to resort to unreliable heuristics ("if you haven't added it back after 10ms you probably don't need it").

An alternative approach would be to provide a way for people to denote that they intend to re-use a particular custom element and therefore the underlying component instance should not be destroyed after the element has been removed from the DOM.
Something like:

```js
myCustomElement.remove();

// I might reuse this in the future. Do not destroy the underlying component.
myCustomElement.ngRetainComponentInstance();
```

As a work-around, you can basically do this manually today (by using some private properties 😇):

```js
myCustomElement.remove();

myCustomElement.ngElementStrategy.scheduledDestroyFn()
myCustomElement.ngElementStrategy.scheduledDestroyFn = null;
```

[1]: https://github.com/angular/angular/blob/1373a98e259b44c28c0547394bbd48643fe81911/packages/elements/src/component-factory-strategy.ts#L172-L173

**Comment by kseamon** [comment]
The race condition I was referring to is that it appears that the strategy factory assumes there is only ever one instance of the element, which would be fine if elements created a separate strategy instance per dom node. Looking at NgElementImpl, though, it appears to cache the strategy instance and reuse the same one each time.

So a bigger needed change is to either create a separate instance of the strategy per dom node or to make the strategy keep track of multiple dom node / component instance pairs via something like a WeakMap.

**Comment by kseamon** [comment]
For the content projection issue, could we just put the projected elements back into the element when we disconnect?

**Comment by kseamon** [comment]
There's another issue as well - when the first instance of the component is destroyed, Angular's animation engine then tries to remove the custom element node. So after the new one is created, we're seeing the animation engine circle back and remove the custom element which destroys the component again.

This was probably never observed before as you wont get this far unless you first null out the scheduledDestroyFn first.

**Comment by kseamon** [comment]
Proposed solution for this one: create a div within the custom element and attach our component to it. This might be a slightly breaking change but I think will fix this issue.

**Comment by yefim** [comment]
After setting `scheduledDestroyFn` to when the component is destroyed and implementing the div fix in the comment above, I also noticed that Angular will not initialize the inputs with the values from the DOM. Neither `preExistingValues` nor `initialInputValues` were set. This led me to finding a related issue https://github.com/angular/angular/issues/30848 about HTML attributes. I modified the connectedCallback hook to always populate `preExistingValues` and to check the DOM attributes (via `this.getAttribute`). This seems to work as expected.

Here's the diff to get that last bit working:
![Screen Shot 2020-09-11 at 12 09 51 AM](https://user-images.githubusercontent.com/1170843/92880675-1eda0b80-f3c3-11ea-961c-688052bdfaa1.png)

The code could definitely be cleaned up and there may be edge cases I'm missing but it achieves my goal of initializing a new component with the correct inputs set. I'm open to any feedback :)

**Comment by gkalpak** [comment]
Thx for the follow-up, @yefim / @kseamon :+1:

I have left some feedback on #38816 (see https://github.com/angular/angular/pull/38816#pullrequestreview-487292105). In a nutshell:

- _Regarding https://github.com/angular/angular/issues/38778#issuecomment-690711582 (putting the projected elements back into the element when disconnected):_

    We could, but it is not easy to know what the projected elements are. For example, if we keep track of the originally projected nodes, they might have been replaced with other nodes (e.g. by a structural directive or manually in a non-Angular app) or other nodes might have been inserted besides them, which were supposed to be retained. Tons of edge cases 😅 

- _Regarding https://github.com/angular/angular/issues/38778#issuecomment-690716057 (attaching the component to a new `<div>` inside the custom element to work around the animation issue):_

    This sounds too breaky :grin: I would rather fix this elsewhere (see https://github.com/angular/angular/pull/38816#discussion_r487508143).

- _Regarding trying to fix component re-instantiation:_

    I am little wary and would prefer to explore other approaches first. See https://github.com/angular/angular/pull/38816#pullrequestreview-487292105 for details.

@yefim / @kseamon: LMK what you think.
Also, out of curiosity, does the work-around mentioned at the bottom of https://github.com/angular/angular/issues/38778#issuecomment-690433315 work for you?

**Comment by yefim** [comment]
The work-around at the bottom of https://github.com/angular/angular/issues/38778#issuecomment-690433315 does not work for us for 2 reasons: (1) the component is destroyed by the `@angular/animations` package (I've filed https://github.com/angular/angular/issues/38962 for that) and (2) the inputs are not reinitialized from the HTML attributes. Furthermore, we would prefer to not couple Angular Element removal logic with our host application.

My main goal when filing this bug was to see if we could get removing and re-adding an Angular Element to work. I don't mind if the element needs to re-render or reinitialize. I solely want the host application to be capable of removing and re-adding the element and the element to be in the correct state when re-added.

**Comment by gkalpak** [comment]
@yafim, would it be possible to create a minimal reproduction showing your usecase (on [StackBlitz][1] or something similar) or failing unit test?

It would help make the discussion less theoretical and it would make it easier to try out different solutions :grin:

[1]: https://stackblitz.com/

**Comment by gkalpak** [comment]
FWIW, I played around with the idea of detaching the component from the `ApplicationRef` (instead of destroying it) when the element is disconnected and re-attaching it when the element is connected (for reference: [internal discussion][1]).

This seems to work fine - i.e. the component is kept intact and can be moved around DOM without issues, but it is garbage-collected as expected when all references are dropped. Here is my POC: [StackBlitz project][2]

However, one issue is that (since the component is not destroyed) any destroy hooks (such as `ngOnDestroy()`) will not be called. This is quite problematic when there is clean-up that needs to be performed (such as canceling intervals, etc.) 😞

[1]: https://angular-team.slack.com/archives/C040TF8UT/p1600071502027800
[2]: https://stackblitz.com/edit/ngelements-issue-38778?file=src%2Fapp%2Fapp.component.ts

**Comment by Oni07R** [comment]
> ```js
> scheduledDestroyFn = null
> ```

Big thank you!! The scheduledDestroyFn = null workaround saved me! :D

**Comment by MaxArt2501** [comment]
This bug is so severe that basically prevents Angular Elements from being used for developing a component library, unless developers are asked to avoid detaching and reattaching elements - which is a totally ludicrous request.

I've fiddled around this problem using a custom `NgElementStrategyFactory` class, basically copying the original `ComponentNgElementStrategyFactory` (this is not good DX, folks... I hope it'll get there) but removing the `return` in the `connect` method (also references to Zone.js). That seemed fine and the component would get recreated, but... then it crashes as soon as `ShadowDomRenderer` [tries to reattach the Shadow DOM](https://github.com/angular/angular/blob/71e14a71f5b8e0a7569df8442402a0c122fec870/packages/platform-browser/src/dom/dom_renderer.ts#L321) without checking if there's one already (note: this is also [incompatible with Declarative Shadow DOM](https://web.dev/declarative-shadow-dom/#hydration)) and eventually reusing or removing its content.

I don't know if it's worth considering making a PR out of this, since many other things are out of my comprehension of the problem, but let me know. I haven't even checked if the other encapsulation modes are fine.

Anyway, I'd like to say that it's natural to compare the life cycle of Web Components to the one of Angular components, so that `ngOnInit`/`ngAfterViewInit` are basically the equivalent of `connectedCallback`, while `ngOnDestroy` is basically `disconnectedCallback`. As developers, I think we need to rely on this.

**Comment by aetzlecx** [comment]
Is it possible that my issue described here https://stackoverflow.com/questions/69074888/angular-ng-if-directive-breaks-web-component-exported-with-angular-elements is related to this issue here?

**Comment by aetzlecx** [comment]
As @MaxArt2501 already stated it's a major bug and it's also a huge risk for us and our client since we're not able to provide a proper solution to export a component library. It would be cool if the community could give at least feedback if this is a major bug and needs a proper solution or if it's just a tiny thing.

While we're waiting for this we we discovered the following fix:

We basically take the outcome of `createCustomElement` function and wrap the constructor and do the following:
* `connectedCallback` was overwritten to be executed only once and properly initialize the component
* `disconnectedCallback` was overwritten entirely to do nothing
* Another method `destroyComponent` was introduced to destroy the component manually (by calling the `disconnectedCallback` of `NgElementImpl`)

The upsides are:
* This gives the developer the flexibility of adding/removing components to the DOM without loosing the state ...
* ...and let them also destroy the component when no longer needed.
* I had not to touch Angular/Elements code

The downsides are:
* Extra step needed to destroy component (removing element from the DOM is not enough)
* When the component is not attached to the DOM it's still somehow alive (e.g. listening to events) which could cause performance degredations

Is there somebody in the community to inform us if there is something better than this workaround?

Any help is really appreciated
Many thanks

**Comment by gkalpak** [comment]
>> The problem is that it will also not work as expected in other situations and would be a significant breaking change
>
> Can you please elaborate on that what you mean? At the moment we're having this workaround deployed for our customer to build a component library based on exported angular components as web components and we're not facing any issues so far. Maybe we've overlooked something and your insight could help us.

With your suggestion, you will have to manually call `destroyComponent()` for the component to be properly cleaned up. If we make this change, tit can potentially break existing apps by failing to perform clean-up when components are removed, unless people update their code to manually call `destroyComponent()` (thus it is a breaking change).

Furthermore, it is not always necessary to manually call `destroyComponent()` for several reasons:
- You might not control the removal of nodes from the DOM. (For example, if you use another framework/library that manages DOM operations.)
- You might not know whether the user will use the component again. (For example, imagine a toggle button that shows/hides a component.)

Does that make sense?

##
>> Implementing your own ElementStrategy is a reasonable work-around, though.
>
> This is what our workaround is based on, however we need some hacky tricks to do it as we're not able to access all the code from `@angular/material`.
>
> At the moment inside `createCustomElement` a `NgElementStrategyFactory` can be passed that can create `NgElementStrategy` instances for each component. Concrete implementations are `ComponentNgElementStrategyFactory` and `ComponentNgElementStrategy` which I'm not able to overwrite since they're not exported via [public_api.ts](https://github.com/angular/angular/blob/master/packages/elements/public_api.ts).
>
> What would speak against to expose these concrete implementation's so that we could build our workaround based on them and we just overwrite methods we need?

The problem with making `ComponentNgElementStrategy[Factory]` public API is that it would make it much harder for us to change them (since they would be subject to our [deprecation/support policies][1]). Thus, (for the few people that might need to provide their own strategy) we deem it acceptable to duplicate the parts of the code that you need.

Maybe we could revisit that decision...

[1]: https://angular.io/guide/releases#deprecation-practices

**Comment by DanielIcc** [comment]
Hi i have the same problem adding to angular js legacy app, ng-if breaks the angular web component.

**Comment by cumminsjp** [comment]
I'm seeing this same issue in Angular 14

**Comment by jecarney** [comment]
Any movement on this issue? In our issue the problem comes up when we renav into the web component (which should go from a web component deep link to the web component home page) and the web component becomes unresponsive.

**Comment by HazzMan2409** [comment]
Bumped into this one today. In addition to the code posted for the [workaround](https://github.com/angular/angular/issues/38778#issuecomment-690433315)  I also found that I needed to set the component reference to null:

```typescript
myCustomElement.remove();

myCustomElement.ngElementStrategy.scheduledDestroyFn();
myCustomElement.ngElementStrategy.scheduledDestroyFn = null;
myCustomElement.ngElementStrategy.componentRef = null
```

**Comment by RazeiXello** [comment]
I also bumped into this recently, we found that using a [setter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/set) seems to give consistent results when trying to assign values from inputs on any subsequent initializations of an angular/elements web component.

```ts

private _myValue = '';

@Input()
set myValue(value: string) {
  if (value) {
    this._myValue = value;
  }
}
```

The caveat is that the setters seem to run after the child component loaded in the `router-outlet` tries to access the value, which was solved with a boolean that's set to true in the afterViewInit function.

Something like this in the `AppComponent`:
 
 ```html
<div *ngIf="isInit">
  <router-outlet></router-outlet>
</div>
 ```
 ```ts
 public isInit = false;

// allows the setters to execute
ngAfterViewInit(): void {
  this.isInit = true;
}
```

Doesn't solve the root issue, but has unblocked our work thus far.

**Comment by rahulwakade** [comment]
Can we have a update on this from Angular team?

**Comment by straeger** [comment]
Hello, 
I also faced this issue within a component of our project. It was hard to reproduce. We have the effect that the components where directly destroyed when they where already created and rendered in the DOM. 

After the initial rendering and navigating away within the outer angular app via ng-routing and re route to the component which uses our customElements we had the effect that from time to time ( 1 - 8 times ) the component directly was destroyed and not rendered. 

I also had to wrap the constructor and overwrite the disconnectedCallback (like @aetzlecx did on Sep 14, 2021) to do nothing to get a stable rendering. 

To me it looks like their is still an issue with @angular/elements 17.3.12 and attaching and detaching components to the DOM. 

Unfortunately  I currently have no time to create a stackblitz to reproduce this issue. Maybe I find some time later.

**Comment by jberndsen-planon** [comment]
I can confirm we face these precise challenges still in Angular 19.
